build.gradle.kts

```diff
@@ -160,14 +317,33 @@ tasks {
         }
     }
 
-    test {
+    fun Test.applyCommonTestConfig() {
+        group = "verification"
         maxHeapSize = "1g" // When this line was added Xmx 512m was the default, and we saw OOMs
         maxParallelForks = Math.max(1, Runtime.getRuntime().availableProcessors() / 2)
         useJUnitPlatform()
         // report is always generated after tests run
         finalizedBy(jacocoTestReport)
     }
 
+    test {
+        applyCommonTestConfig()
+    }
+
+    /** Runs the JUnit test on the minified jar. */
+    register<Test>("minifyTest") {
+        applyCommonTestConfig()
+        classpath = project.configurations.testRuntimeClasspath.get() + project.sourceSets.test.get().output + minifyJar.outputs.files
+        dependsOn(minifyJar)
+    }
+
+    /** Runs the JUnit test on the shadow jar. */
+    register<Test>("shadowTest") {
+        applyCommonTestConfig()
+        classpath = project.configurations.testRuntimeClasspath.get() + project.sourceSets.test.get().output + shadowJar.get().outputs.files
+        dependsOn(minifyJar)
+    }
```

Comment:
Which tests does this run? ion-java's tests? Serially or in parallel?

Reply:
Yes, this runs Ion Java's tests.

They don't run as a dependency of the `build` task or any other task. You have to manually run them. They _can_ run in parallel, but whether they do depends on the rest of the task graph and the order that Gradle chooses to execute tasks that are independent of each other.

I don't think there's any need to have `shadowTest` as a dependency of the `build` task, since it's testing an intermediate step (the shaded, but not minified jar); it can be useful though for debugging where something went wrong if `minifyTest` fails.

I would like to have `minifyTest` get run automatically somehow, but `publish` might be too late, and I don't want to add it to `build` because it _will_ slow things down (even if it does run in parallel). How would you feel if I added it to the GitHub Actions workflow?

Reply:
Adding it to the workflow sounds good to me.